pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixInstanceDataManager.java

```diff
@@ -212,34 +217,53 @@ public void reloadSegment(String tableNameWithType, String segmentName, boolean
   }
 
   @Override
-  public void reloadAllSegments(String tableNameWithType, boolean forceDownload) {
+  public void reloadAllSegments(String tableNameWithType, boolean forceDownload, int parallelism)
+      throws Exception {
     LOGGER.info("Reloading all segments in table: {}", tableNameWithType);
     TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
     Preconditions.checkNotNull(tableConfig);
 
     Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, tableNameWithType);
 
     List<String> failedSegments = new ArrayList<>();
-    Exception sampleException = null;
+
     List<SegmentMetadata> segmentsMetadata = getAllSegmentsMetadata(tableNameWithType);
+
+    int nosWorkers = Math.min(segmentsMetadata.size(), parallelism);
+
+    ExecutorService workers = Executors.newFixedThreadPool(nosWorkers);
+
+    final AtomicReference<Exception> sampleException = new AtomicReference<>();
+
+    CountDownLatch latch = new CountDownLatch(nosWorkers);
+
     for (SegmentMetadata segmentMetadata : segmentsMetadata) {
-      try {
-        reloadSegment(tableNameWithType, segmentMetadata, tableConfig, schema, forceDownload);
-      } catch (Exception e) {
-        String segmentName = segmentMetadata.getName();
-        LOGGER.error("Caught exception while reloading segment: {} in table: {}", segmentName, tableNameWithType, e);
-        failedSegments.add(segmentName);
-        sampleException = e;
-      }
+      CompletableFuture.runAsync(() -> {
+        try {
+          reloadSegment(tableNameWithType, segmentMetadata, tableConfig, schema, forceDownload);
+        } catch (Exception e) {
+          String segmentName = segmentMetadata.getName();
+          LOGGER.error("Caught exception while reloading segment: {} in table: {}", segmentName, tableNameWithType, e);
+          failedSegments.add(segmentName);
+          sampleException.set(e);
+        } finally {
+          latch.countDown();
+        }
+      }, workers);
```

Comment:
I think you can make better use of the `CompletableFuture` API here:

```java
CompletableFuture.allOf(segmentsMetadata.stream()
.map(segmentMetadata -> CompletableFuture.runAsync(() -> {... reload the segment ...}, workers)
.toArray(CompletableFuture[]::new))
.get(timeout, TimeUnit.MILLISECONDS);
```

Which would eliminate the latch and make the code easier to reason about.

Reply:
This is excellent. Thanks! Definitely need to revisit `CompletableFuture`'s API once.